backoff.go

```diff
@@ -0,0 +1,68 @@
+package grpc
+
+import (
+	"math/rand"
+	"time"
+)
+
+// DefaultBackoffConfig uses values specified for backoff in
+// https://github.com/grpc/grpc/blob/e1d8d4dc1bdf54273b7ed9d4a495789430cbdf6c/doc/connection-backoff.md.
+var (
+	DefaultBackoffConfig = &BackoffConfig{
+		MaxDelay:  120 * time.Second,
+		baseDelay: 1.0 * time.Second,
+		factor:    1.6,
+		jitter:    0.2,
+	}
+)
+
+// backoffStrategy defines the methodology for backing off after a grpc
+// connection failure.
+//
+// This is unexported until the GRPC project decides whether or not to allow
+// alternative backoff strategies. Once a decision is made, this type and its
+// method may be exported.
+type backoffStrategy interface {
+	// Backoff returns the amount of time to wait before the next retry given
+	// the number of consecutive failures.
+	backoff(retries int) time.Duration
+}
+
+// BackoffConfig defines the parameters for the default GRPC backoff strategy.
+type BackoffConfig struct {
+	// MaxDelay is the upper bound of backoff delay.
+	MaxDelay time.Duration
+
+	// TODO(stevvooe): The following fields are not exported, as allowing changes
+
+	// baseDelay is the amount of time to wait before retrying after the first
+	// failure.
+	baseDelay time.Duration
+
+	// Factor is applied to the backoff after each retry.
+	factor float64
+
+	// jitter provides a range to randomize backoff delays.
+	jitter float64
+}
+
+func (bc *BackoffConfig) backoff(retries int) (t time.Duration) {
```

Comment:
Yeah, pointer is type-safer though.

On Wed, Mar 30, 2016 at 8:08 PM, Stephen Day <notifications@github.com>
wrote:

> In backoff.go
> <https://github.com/grpc/grpc-go/pull/601#discussion_r57984147>:
>
> > +	MaxDelay time.Duration
> > +
> > +	// TODO(stevvooe): The following fields are not exported, as allowing changes
> > +
> > +	// baseDelay is the amount of time to wait before retrying after the first
> > +	// failure.
> > +	baseDelay time.Duration
> > +
> > +	// Factor is applied to the backoff after each retry.
> > +	factor float64
> > +
> > +	// jitter provides a range to randomize backoff delays.
> > +	jitter float64
> > +}
> > +
> > +func (bc *BackoffConfig) backoff(retries int) (t time.Duration) {
>
> We can change to use a value here. The interface will work correctly
> either way.
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/grpc/grpc-go/pull/601/files/cfa10b2c554ed1f23df44716d535188b499a1490#r57984147>
>